src/garage/tf/algos/rl2npo.py

```diff
@@ -0,0 +1,188 @@
+            'no_entropy'. The type of entropy method to use. 'max' adds the
+            dense entropy to the reward for each time step. 'regularized' adds
+            the mean entropy to the surrogate objective. See
+            https://arxiv.org/abs/1805.00909 for more details.
+        flatten_input (bool): Whether to flatten input along the observation
+            dimension. If True, for example, an observation with shape (2, 4)
+            will be flattened to 8.
+        fit_baseline (str): Either 'before' or 'after'. See above docstring for
+            a more detail explanation. Currently it only supports 'before'.
+        name (str): The name of the algorithm.
+
+    Raises:
+        ValueError: When argument 'fit_baseline' != 'before'.
+
+    """
+
+    def __init__(self,
+                 env_spec,
+                 policy,
+                 baseline,
+                 scope=None,
+                 max_path_length=500,
+                 discount=0.99,
+                 gae_lambda=1,
+                 center_adv=True,
+                 positive_adv=False,
+                 fixed_horizon=False,
+                 pg_loss='surrogate',
+                 lr_clip_range=0.01,
+                 max_kl_step=0.01,
+                 optimizer=None,
+                 optimizer_args=None,
+                 policy_ent_coeff=0.0,
+                 use_softplus_entropy=False,
+                 use_neg_logli_entropy=False,
+                 stop_entropy_gradient=False,
+                 entropy_method='no_entropy',
+                 flatten_input=True,
+                 fit_baseline='before',
+                 name='NPO'):
+        self._fit_baseline = fit_baseline
+        if fit_baseline != 'before':
+            raise ValueError('RL2NPO only supports fit baseline before')
+        super().__init__(env_spec=env_spec,
+                         policy=policy,
+                         baseline=baseline,
+                         scope=scope,
+                         max_path_length=max_path_length,
+                         discount=discount,
+                         gae_lambda=gae_lambda,
+                         center_adv=center_adv,
+                         positive_adv=positive_adv,
+                         fixed_horizon=fixed_horizon,
+                         pg_loss=pg_loss,
+                         lr_clip_range=lr_clip_range,
+                         max_kl_step=max_kl_step,
+                         optimizer=optimizer,
+                         optimizer_args=optimizer_args,
+                         policy_ent_coeff=policy_ent_coeff,
+                         use_softplus_entropy=use_softplus_entropy,
+                         use_neg_logli_entropy=use_neg_logli_entropy,
+                         stop_entropy_gradient=stop_entropy_gradient,
+                         entropy_method=entropy_method,
+                         flatten_input=flatten_input)
+
+    def optimize_policy(self, itr, samples_data):
```

Comment:
how does this differ from optimize_policy in regular NPO?

Reply:
This fit baseline first, regular NPO fit baseline after

Reply:
Wait, then why is there a constructor argument which tells NPO when to fit the baseline? It seems like this is a mix of 2 designs

1. Modify NPO to support some weird things (like fitting the baseline before), then pass only constructor params
1. Make special versions of NPO which do what you want.

I'm kind of in favor of 1, but I don't mind either as long as you are consistent.